### PR TITLE
feat: Improve `ScoreAnalysis` debug information

### PR DESCRIPTION
**Pull request by zepfred**
This pull request adds a summary logic to the `ScoreAnalysis` class and addresses minor issues.

### CODE REVIEW

File: core/src/main/java/ai/timefold/solver/core/api/score/analysis/ConstraintAnalysis.java [hunk 1]
```diff
@@ -56,6 +60,16 @@ static <Score_ extends Score<Score_>> ConstraintAnalysis<Score_> of(ConstraintRe
         Objects.requireNonNull(score);
     }
 
+    public int matchCount() {
```
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/api/score/analysis/ConstraintAnalysis.java:63]: Deserves a Javadoc to explain the exception maybe being thrown.

Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/api/score/analysis/ConstraintAnalysis.java:63]: Deserves a Javadoc to explain the exception maybe being thrown.

File: core/src/main/java/ai/timefold/solver/core/api/score/analysis/ConstraintAnalysis.java [hunk 2]
```diff
@@ -156,6 +170,51 @@ public String constraintName() {
         return constraintRef.constraintName();
     }
 
+    /**
+     * Returns a diagnostic text that explains part of the score quality through the {@link ConstraintAnalysis} API.
```
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/api/score/analysis/ConstraintAnalysis.java:174]:
```suggestion
     * Returns a diagnostic text that explains part of the score quality through the {@link ConstraintAnalysis} API. 
     * The string is built fresh every time the method is called.
```

File: core/src/main/java/ai/timefold/solver/core/api/score/analysis/ConstraintAnalysis.java [hunk 3]
```diff
@@ -156,6 +170,51 @@ public String constraintName() {
...
+     *
+     * @return never null
+     */
+    public String summarize() {
+        StringBuilder summary = new StringBuilder();
```
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/api/score/analysis/ConstraintAnalysis.java:179]: In new code, we use `var` where possible.

File: core/src/main/java/ai/timefold/solver/core/api/score/analysis/ConstraintAnalysis.java [hunk 4]
```diff
@@ -156,6 +170,51 @@ public String constraintName() {
...
+        if (constraintMatches == null) {
+            throw new IllegalArgumentException("""
+                    The constraint matches must be non-null.
+                    Maybe use ScoreAnalysisFetchPolicy.FETCH_ALL to request the score analysis
+                    """);
+        }
+        if (constraintMatches.isEmpty()) {
+            summary.append("""
+                            %s: constraint (%s) has no matches.
+                    """.formatted(score().toShortString(), constraintRef().constraintName()));
```
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/api/score/analysis/ConstraintAnalysis.java:196]: Arguably, for short single-line strings, ```"""``` is unnecessary and more verbose.
But careful to maintain the newline at the end.

File: core/src/main/java/ai/timefold/solver/core/api/score/analysis/ConstraintAnalysis.java [hunk 5]
```diff
@@ -156,6 +170,51 @@ public String constraintName() {
...
+        } else {
+            summary.append("""
```
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/api/score/analysis/ConstraintAnalysis.java:198]: Dtto.

File: core/src/main/java/ai/timefold/solver/core/api/score/analysis/ConstraintAnalysis.java [hunk 6]
```diff
@@ -156,6 +170,51 @@ public String constraintName() {
...
+                            %s: constraint (%s) has %s matches:
+                    """.formatted(score().toShortString(), constraintRef().constraintName(), constraintMatches.size()));
+        }
+        constraintMatches.stream()
+                .sorted(matchScoreComparator)
+                .limit(DEFAULT_SUMMARY_CONSTRAINT_MATCH_LIMIT)
+                .forEach(match -> summary.append("""
```
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/api/score/analysis/ConstraintAnalysis.java:205]: Dtto.

File: core/src/main/java/ai/timefold/solver/core/api/score/analysis/ConstraintAnalysis.java [hunk 7]
```diff
@@ -156,6 +170,51 @@ public String constraintName() {
...
+                                    %s: justified with (%s)
+                        """.formatted(match.score().toShortString(),
+                        match.justification())));
+        if (constraintMatches.size() > DEFAULT_SUMMARY_CONSTRAINT_MATCH_LIMIT) {
+            summary.append("""
```
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/api/score/analysis/ConstraintAnalysis.java:210]: Dtto.

File: core/src/main/java/ai/timefold/solver/core/api/score/analysis/ScoreAnalysis.java [hunk 8]
```diff
@@ -141,4 +145,63 @@ public Collection<ConstraintAnalysis<Score_>> constraintAnalyses() {
         return constraintMap.values();
     }
 
+    /**
+     * Returns a diagnostic text that explains the solution through the {@link ConstraintAnalysis} API to identify which
+     * constraints cause that score quality.
```
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/api/score/analysis/ScoreAnalysis.java:150]:
```suggestion
     * constraints cause that score quality.
     * The string is built fresh every time the method is called.
```

File: core/src/main/java/ai/timefold/solver/core/api/score/analysis/ScoreAnalysis.java [hunk 9]
```diff
@@ -141,4 +145,63 @@ public Collection<ConstraintAnalysis<Score_>> constraintAnalyses() {
...
+                .forEach(constraint -> {
+                    var matches = constraint.matches();
+                    if (matches == null) {
+                        throw new IllegalArgumentException("""
+                                The constraint matches must be non-null.
+                                Maybe use ScoreAnalysisFetchPolicy.FETCH_ALL to request the score analysis
+                                """);
+                    }
+                    if (matches.isEmpty()) {
+                        summary.append("""
```
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/api/score/analysis/ScoreAnalysis.java:182]: One-line string, no need for string literals.

File: core/src/main/java/ai/timefold/solver/core/api/score/analysis/ScoreAnalysis.java [hunk 10]
```diff
@@ -141,4 +145,63 @@ public Collection<ConstraintAnalysis<Score_>> constraintAnalyses() {
...
+                                        %s: constraint (%s) has no matches.
+                                """.formatted(constraint.score().toShortString(), constraint.constraintRef().constraintName()));
+                    } else {
+                        summary.append("""
+                                        %s: constraint (%s) has %s matches:
```
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/api/score/analysis/ScoreAnalysis.java:187]: Dtto.

File: core/src/main/java/ai/timefold/solver/core/api/score/analysis/ScoreAnalysis.java [hunk 11]
```diff
@@ -141,4 +145,63 @@ public Collection<ConstraintAnalysis<Score_>> constraintAnalyses() {
...
+                                """.formatted(constraint.score().toShortString(), constraint.constraintRef().constraintName(),
+                                matches.size()));
+                    }
+                    matches.stream()
+                            .sorted(matchScoreComparator)
+                            .limit(DEFAULT_SUMMARY_CONSTRAINT_MATCH_LIMIT)
+                            .forEach(match -> summary.append("""
```
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/api/score/analysis/ScoreAnalysis.java:194]: Dtto.

File: core/src/main/java/ai/timefold/solver/core/api/score/analysis/ScoreAnalysis.java [hunk 12]
```diff
@@ -141,4 +145,63 @@ public Collection<ConstraintAnalysis<Score_>> constraintAnalyses() {
...
+                                                %s: justified with (%s)
+                                    """.formatted(match.score().toShortString(),
+                                    match.justification())));
+                    if (matches.size() > DEFAULT_SUMMARY_CONSTRAINT_MATCH_LIMIT) {
+                        summary.append("""
```
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/api/score/analysis/ScoreAnalysis.java:199]: Dtto.

File: core/src/main/java/ai/timefold/solver/core/api/score/analysis/ScoreAnalysis.java [hunk 13]
```diff
@@ -53,6 +55,8 @@
 public record ScoreAnalysis<Score_ extends Score<Score_>>(Score_ score,
```
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/api/score/analysis/ScoreAnalysis.java:55]: Arguably, `ScoreAnalysis` needs an informative `toString()` too.
Something like `Score analysis of score X with Y constraints`.

Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/api/score/analysis/ScoreAnalysis.java:55]: Also, do we want to make it simple to figure out that the score was not initialized?
To find out today, you have to parse the score and check for `init`.
Maybe we want to provide a new field which does this on behalf of the user?

**Review by triceo — CHANGES_REQUESTED** [review]
Good start!
Please double-check serialization of the analysis in `persistence-jackson`. Nothing should have changed.

**Comment by triceo** [comment]
Also, it makes sense to fix [this](https://sonarcloud.io/project/issues?resolved=false&sinceLeakPeriod=true&pullRequest=923&id=ai.timefold%3Atimefold-solver&open=AZBbJHPn5hyUaEYfs9xR).

**Review by triceo — APPROVED** [review]
There are some new Sonar warnings which I think are good to address.
Other than that, LGTM.

**Comment by zepfred** [comment]
> There are some new Sonar warnings which I think are good to address. Other than that, LGTM.

I remember encountering errors when using `%n` on Mac. Let me attempt it again.

**Comment by zepfred** [comment]
> > There are some new Sonar warnings which I think are good to address. Other than that, LGTM.
> 
> I remember encountering errors when using `%n` on Mac. Let me attempt it again.

Here is the problem:

https://github.com/TimefoldAI/timefold-solver/actions/runs/9750701051/job/26910684905?pr=923#step:4:2824

**Pull request merged**